smartsim/_core/utils/serialize.py

```diff
@@ -0,0 +1,237 @@
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+from __future__ import annotations
+
+import json
+import time
+import typing as t
+from pathlib import Path
+
+import smartsim._core._cli.utils as _utils
+from smartsim._core.config import CONFIG
+
+if t.TYPE_CHECKING:
+    from smartsim import Experiment
+    from smartsim._core.control.manifest import LaunchedManifest as _Manifest
+    from smartsim.database.orchestrator import Orchestrator
+    from smartsim.entity import DBNode, Ensemble, Model
+    from smartsim.entity.dbobject import DBModel, DBScript
+    from smartsim.settings.base import BatchSettings, RunSettings
+
+
+TStepLaunchMetaData = t.Tuple[
+    t.Optional[str], t.Optional[str], t.Optional[bool], str, str, Path
+]
+TELMON_SUBDIR: t.Final[str] = ".smartsim/telemetry"
+MANIFEST_FILENAME: t.Final[str] = "manifest.json"
+
+
+def save_launch_manifest(manifest: _Manifest[TStepLaunchMetaData]) -> None:
+    if not CONFIG.telemetry_enabled:
+        return
+
+    manifest.metadata.run_telemetry_subdirectory.mkdir(parents=True, exist_ok=True)
+
+    new_run = {
+        "run_id": manifest.metadata.run_id,
+        "timestamp": int(time.time_ns()),
+        "model": [
+            _dictify_model(model, *telemetry_metadata)
+            for model, telemetry_metadata in manifest.models
+        ],
+        "orchestrator": [
+            _dictify_db(db, nodes_info) for db, nodes_info in manifest.databases
+        ],
+        "ensemble": [
+            _dictify_ensemble(ens, member_info)
+            for ens, member_info in manifest.ensembles
+        ],
+    }
+    try:
+        with open(manifest.metadata.manifest_file_path, "r", encoding="utf-8") as file:
+            manifest_dict = json.load(file)
+    except (FileNotFoundError, json.JSONDecodeError):
+        manifest_dict = {
+            "schema info": {
+                "schema_name": "entity manifest",
+                "version": "0.0.1",
+            },
+            "experiment": {
+                "name": manifest.metadata.exp_name,
+                "path": manifest.metadata.exp_path,
+                "launcher": manifest.metadata.launcher_name,
+            },
+            "runs": [new_run],
+        }
+    else:
+        manifest_dict["runs"].append(new_run)
+    finally:
+        with open(manifest.metadata.manifest_file_path, "w", encoding="utf-8") as file:
+            json.dump(manifest_dict, file, indent=2)
+
+
+def _dictify_model(
```

Comment:
Wouldn't these dictify methods be more appropriate to add to `__dict__` on each entity's class definition?

Reply:
I would actually argue they should not be instance methods on the `Model`/`Ensemble`/etc. The reason for this is I didn't want to clutter up the entities with a method that is dedicated specifically for serialization. For instance, I could see a another feature in SmartSim eventually needing to cast entities to dictionaries that are either subsets or superset of these dictionaries and I didn't want to start the pattern of making methods `Model.to_dict_for_serialization()`, `Model.to_dict_for_other_feature()`, etc.

Reply:
In other words, my goal was to try and encapsulate all logic regarding serialization into dedicated a `serialization` module. If these fns are not used outside of this module I didn't see a reason to expose them more than necessary.

Let me know if you think this seems justified, but if not I would be happy to revisit this decision!

Reply:
Punted this into a different ticket